Realm native token and tipping

aligatorr89 opened the pull request:
Realm native token and tipping

- add native token and more tokens to realm
- tip with realm token sand show it on Taggr posts
- e2e tests, rust unit tests
- integrate IcExplorer for auto discovery of tokens in wallet - UX improvement

Workflow for tips:
1. tip client -> external ledger
2. add to the view optimistically (good UX and quick)
3. verify tip with `icrc3_get_blocks` and add it to post, cost is 5 credits to cover for canister->canister call

For next time - min_native_token_balance:
- provide a funding model for DAO - canister -> canister calls burn cycles
- we can increase post costs to fund it
- if below min, we remove post -> more credits burned to disinsentivize

GEMINI review:
[taggr-review.md](https://github.com/user-attachments/files/22083666/taggr-review.md)

Anugrahtaggr commented:
.

aligatorr89 committed:
Realm native token and tipping

Fix ICRC structs for decoding

Custom popup and tip flow with it

Optimistic tip view

Tips and tokens formatting

Filter non icrc from icp-swap

Stable memory tips

New staging dev2

e2e tests

staging dev2 links

performance fixes

Remove native token and after rebase fixes

logo fallback

review comments

taggrx commented on Makefile:
```diff
@@ -6,6 +6,11 @@ staging_deploy:
 	FEATURES=staging dfx build
 	FEATURES=staging dfx --identity prod deploy --network staging taggr
 
+staging_dev2_deploy:
```
Instead of adding a new command, let's parameterize the existing `staging_deploy` like how I parameterized `make release`. E.g. I'd rather suggest to deploy to the selected staging by something like:

CANISTER=staging_dev2 make staging_deploy

taggrx replied:
Another comment I have is what does `dev` in the name of the second staging environment imply? Staging is obviously dev, not prod. How about simply `staging2`?

aligatorr89 replied:
done, original staging is default

taggrx commented on e2e/test1.spec.ts:
```diff
@@ -1,11 +1,45 @@
+                resolve(null);
+            }
+        });
+        // Click button after lisenter is setup
```
```suggestion
        // Click button after listener is setup
```

taggrx commented on e2e/test1.spec.ts:
```diff
@@ -265,6 +274,117 @@ test.describe("Upgrades & token transfer flow", () => {
         ).toEqual(2);
     });
+
+    test.describe("IC TOKENS", () => {
```
Awesome! Thank you!

aligatorr89 replied:
awesome!

taggrx commented on e2e/test3.spec.ts:
```diff
@@ -104,4 +104,128 @@ test.describe("Regular users flow, part two", () => {
     });
+
+    test.describe("Tips", () => {
+        test('Logout and login with "1" user', async () => {
```
Why not using a random name? I had to read the code to understand what "1" user means...

taggrx commented on src/backend/env/mod.rs:
```diff
@@ -145,6 +147,11 @@ pub struct Realm {
     pub posts: Vec<PostId>,
     pub adult_content: bool,
     pub comments_filtering: bool,
+    /// Native token of realm - 1 use case is min balance to post
```
I would describe use cases etc in the white paper.

```suggestion
    /// Native token of the realm
```

aligatorr89 replied:
removed native token due to reasons described

taggrx commented on src/backend/env/mod.rs:
```diff
@@ -778,6 +792,9 @@ impl State {
         if !logo.is_empty() {
             realm.logo = logo;
         }
+        if tokens.clone().map_or(false, |t| t.len() > 50) {
```
We shouldn't clone during the validation, `.as_ref()` should be fine here.

taggrx commented on src/backend/env/mod.rs:
```diff
@@ -778,6 +792,9 @@ impl State {
             realm.logo = logo;
         }
+        if tokens.clone().map_or(false, |t| t.len() > 50) {
+            return Err("tokens count below 50".into());
```
```suggestion
            return Err("tokens count above 50".into());
```

taggrx commented on src/backend/env/post.rs:
```diff
@@ -91,6 +91,9 @@ pub struct Post {
     pub encrypted: bool,
+
+    #[serde(default)]
+    pub has_external_tip: Option<bool>,
```
Please add comments explaining what it is. For example, I didn't see the rest of the code, but seeing this with this name, I have no idea why this field is needed.

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,385 @@
+
+    canister_id: Principal,
+    amount: u128,
+    /// Index from external transaction
```
Why is it needed?

aligatorr89 replied:
- to avoid repeating same tip, only way to make sure it's not repeated
- generally it's good idea to keep externalId

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,385 @@
+    from_principal: Principal,
+    index: u64,
+) -> Result<Tip, String> {
+    let post = Post::get(state, &post_id).ok_or("post not found")?;
```
Nit: reading the post is the most expensive operation here, I'd do it after we know we have all user accounts, so move it down as much as possible

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,385 @@
+        .get_mut(&receiver_id)
+        .expect("user not found")
+        .notify_about_post(
+            format!("@{} tipped you with for your post", sender_name),
```
Something is missing in the message

aligatorr89 replied:
We don't know symbol and decimals in backend. User can still see it in notification by clicking 

![image](https://github.com/user-attachments/assets/6d39a171-bf84-4148-ac2e-c97c0cc9d156)

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,385 @@
+    Ok(tip)
+}
+
+fn convert_memo_to_u64(memo: Option<Vec<u8>>) -> Result<PostId, String> {
```
Is this really a `PostId`?

aligatorr89 replied:
u64 obviouselly :D

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,385 @@
+        num = (num << 8) | byte as u64; // Shift left and add byte
+    }
+
+    Ok(num)
```
I didn't test it, but this is more idiomatic, please TAL:
```suggestion
let mut padded = [0u8; 8];
let len = std::cmp::min(memo_value.len(), 8);
padded[8-len..].copy_from_slice(&memo_value[..len]);
Ok(u64::from_be_bytes(padded))
```

aligatorr89 replied:
amazing, works the same. How did you spot this ?

taggrx commented on src/backend/env/canisters.rs:
```diff
@@ -251,6 +289,20 @@ pub async fn coins_for_one_xdr(coin: &str) -> Result<u64, String> {
+    args: GetTransactionsArgs,
+) -> Result<GetTransactionsResult, String> {
+    let (response,): (GetTransactionsResult,) =
+        call_canister(canister_id, "get_transactions", (args,))
```
But that's not an ICRC standard.

aligatorr89 replied:
It's Rosetta standard, all ledgers seem to have it. It holds only recent transactions by index, excatly what we need here. Thanks for spotting, renamed to `get_transactions

taggrx replied:
But it's not part of any standard agreed by the community for using in canisters. E.g. this wouldn't work for Taggr itself. Why not simply using `icrc3_get_blocks`?

aligatorr89 replied:
It looks interesting, but `to` and `from` are in some weird format . Any way to convert that?

taggrx replied:
`icrc3_get_blocks` doesn't have any `from` or `to`, I'm not sure what you mean.

aligatorr89 replied:
added icrc3_get_blocks

taggrx commented on src/backend/env/mod.rs:
```diff
@@ -145,6 +147,8 @@ pub struct Realm {
     pub adult_content: bool,
     pub comments_filtering: bool,
+    /// Tokens allowed to appear in realm like tips
+    pub tokens: Option<BTreeSet<Principal>>,
```
Did we say we do not white list tip tokens?

aligatorr89 replied:
realm controllers want this, since it gives just few options they want

taggrx commented on src/backend/env/mod.rs:
```diff
@@ -262,6 +266,10 @@ pub struct State {
+
+    // Map of post id to sorted of tip ids and external index
+    #[serde(default)]
+    pub post_tip_indexes: BTreeMap<PostId, Vec<TipId>>,
```
Why is this data structure needed? Why can't all tip data be on the post itself?

aligatorr89 replied:
Solid design, 1-many table. For Taggr posts, it's wasteful to load when it's never shown unless clicked on post menu.

taggrx replied:
This is wrong: if it is attached to the post, you don't need to store the postID separately and en empty vector is a dangling pointer which is allocates 0 bytes! With the global table, you introduce a whole new structure in the persistent storage and store each tip separately, which is also conceptually weird to explain to others: why storing parts of the posts separately? And since the persistent storage always stores data in blocks, you almost always allocate a couple of bytes that you'll never use. This is what is really wasteful for no good reason, IMO.

aligatorr89 replied:
I wanted to have tips in stable memory. Could solve it by BTreeMap in stable memory - then this heap index can be removed, wdyt?

taggrx replied:
Thanks! It's much cleaner and simpler now (and more space efficient!)

taggrx commented on src/backend/updates.rs:
```diff
@@ -624,6 +627,64 @@ fn cancel_bid() {
 }
 
+#[export_name = "canister_update add_external_icrc_transaction"]
+fn add_external_icrc_transaction() {
```
I would move all actual logic to the tip.rs module, to keep the update handler  tidy.

aligatorr89 replied:
good idea

taggrx commented on src/backend/updates.rs:
```diff
@@ -624,6 +627,64 @@ fn cancel_bid() {
 
+#[export_name = "canister_update add_external_icrc_transaction"]
+fn add_external_icrc_transaction() {
+    let (canister_id_as_str, start_index, post_id): (String, u64, PostId) = parse(&arg_data_raw());
```
I don't understand... Looks like I can call this function with any post_id and _any_ start_index I want (even if I didn't make it) and this will be just added as a tip to that post?!

aligatorr89 replied:
if you want to burn 5 credits for nothing yes :)

taggrx replied:
Wait, but if we think it's ok to attach any untrusted transaction to the post, why do even have any logic in the backend (`tip.rs`)? We could simply make an ICRC transaction on the frontends and simply call this function with all the parameters? Why are we making then an expensive inter-canister call to fetch data that is not verified?

taggrx replied:
Also, I am not sure in general about this conceptually, especially when it matters (for a popular posts) you simply cannot trust if tips are real, or people are trolling others attaching non-exiting large transfers.

taggrx replied:
I thought a bit more, if the tx is done on the frontend using user's principal, why not simply checking the sender of the tx to make sure it was actually issued by the user? (we also need to add the post id to the memo and check that so that the user can't attach one tip to many posts). And then if a check is not successful, penalize users by e.g. draining 100 credits, otherwise I would not charge any credits to not penalize tippers.

aligatorr89 replied:
We verify tx based on memo which holds postId and index - there is no way for double add. We also check sending principal is the same as from param and amount.

taggrx replied:
> We also check sending principal is the same as from param and amount.

Hm, no. `try_tip` is called with `start_index` that returns a transaction, but we never check if this transaction is in any relation with the principal returned from `read(caller)` or am I missing something?

aligatorr89 replied:
added caller check as well

taggrx commented on src/backend/env/canisters.rs:
```diff
@@ -37,6 +38,43 @@ thread_local! {
+pub struct IcrcMemo(pub ByteBuf);
+
+#[derive(CandidType, Clone, Serialize, Deserialize, Debug)]
+pub struct IcrcTransfer {
```
It's the same type as we have in the token.rs module. Let's consolidate?

aligatorr89 replied:
Taggr tokens defintions are slightly different and is hard to re-use due to integration failures, like `u64` instead of `Nat`. It fails during `deserialize` of response which expects slightly different types.

I think we should rather use ICRC types for Taggr tokens definitions. 

P.S. i re-used Account definition now

taggrx replied:
If you'd just switch to the ICRC3 get blocks function, all of this will go away. Let's make progress on the other thread first.

aligatorr89 replied:
Switching to icrc-3 done :)

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,384 @@
+        .ok_or("realm not found")?;
+    let memo_u64 = convert_memo_to_u64(memo)?;
+
+    if post.user != receiver_id {
```
Instead of passing both `to_principal` and `post_id` and then making sure all this matches, let's simply pass `post_id` and extract the receiver from it?

aligatorr89 replied:
Receiver is derived from tx retrieved from external ledger

taggrx replied:
I see now, thanks! What do you think about passing the entire TX into this function and call it `create_post_tip_from_tx`?

Then:

1. The clippy `too_many_arguments` issue goes away
2. The whole logic is encapsulated in one function and not spread over two funcs.

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,384 @@
+        .as_ref()
+        .map_or(false, |tokens| !tokens.contains(&canister_id))
+    {
+        return Err("token is not allowed to tip in realm".to_string());
```
Didn't [we agreed](https://6qfxa-ryaaa-aaaai-qbhsq-cai.icp0.io/#/thread/1432672) we don't want to gateway which tokens the user can use to tip?

aligatorr89 replied:
We can allow yes, but in the FE they will still see those that realm controller have chosen

taggrx replied:
But I thought the whole point of the discussion was to not have it? See my head message of the thread linked above:

> why a realm should have a list of white listed tokens at all?

aligatorr89 replied:
Removed validation

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,384 @@
+    from_principal: Principal,
+    start_index: u64,
+) -> Result<Tip, String> {
+    // DoS protection
```
Please, let's not make tips more unattractive and expensive than they are! Instead let's do this:

- define a local variable with credits penalty for a wrong call (e.g. 20 credits).
- if user has less than needed credits for potential penalty, reject the call.
- fetch the tipping tx
- If the call fails (please don't panic there!) or the tx not found - deduct the penalty.

aligatorr89 replied:
Makes sense. But we should still charge something for successful tip, how much like 5 ?

taggrx replied:
Let's just start with costs of the cheapest reaction?

taggrx commented on src/backend/env/tip.rs:
```diff
@@ -0,0 +1,384 @@
+        }
+    } else {
+        Err(format!(
+            "We could not find transaction at index {}",
```
Who is we? :D 

```suggestion
            "Transaction not found at index {}",
```

taggrx commented on src/frontend/src/common.tsx:
```diff
@@ -472,7 +476,15 @@ export const tokens = (n: number, decimals: number, hideDecimals?: boolean) => {
 
-export const ICP_LEDGER_ID = Principal.fromText("ryjl3-tyaaa-aaaaa-aaaba-cai");
+/** Remove tail zeros */
+export const shortenTokensAmount = (n: number, decimals: number) => {
```
See the `tokens` function in common.tsx, I think this function is not needed.

aligatorr89 replied:
It uses tokens, additionally it removes tailing zeros, like `0.020000` -> `0.02`

taggrx replied:
Ah this is nice! Shouldn't we just make this the default display mode of `tokens` function? Just move it there.

aligatorr89 replied:
seems a bit difficult, don't want to create more test cases :)

taggrx commented on src/frontend/src/common.tsx:
```diff
@@ -1271,3 +1283,250 @@ export const getCanonicalDomain = () =>
+    return canistersFromStorageMap;
+}
+
+export function numberToUint8Array(num: number) {
```
Hm, looks like it's not used anywhere?

aligatorr89 replied:
used in finalizeTip

taggrx commented on src/frontend/src/common.tsx:
```diff
@@ -1271,3 +1283,250 @@ export const getCanonicalDomain = () =>
+    data: any;
+    updated_at: string;
+};
+/** Helper for stale local storage data */
```
Let's stick with one style of comments

taggrx commented on src/frontend/src/common.tsx:
```diff
@@ -1271,3 +1283,250 @@ export const getCanonicalDomain = () =>
+    );
+}
+
+export const Popup = ({
```
Let's not introduce another pop-up implementation, there is `popUp` for that (in common).

aligatorr89 replied:
It is used for simple messages. I'm not sure how to make it work for dynamic React HTML - I don't think it's worth trying to re-use it cause we'll just create too many divergencies.

taggrx commented on src/frontend/src/user_resolve.tsx:
```diff
@@ -3,6 +3,10 @@ import { UserId, UserData, User } from "./types";
 
 export const USER_CACHE: UserData = {};
+export const USER_COMMON_CACHE: Record<
+    number,
```
Is this the `UserId`?

taggrx replied:
Also, why not simply extending the existing `USER_CACHE` and the type `UserData` with what's needed?

aligatorr89 replied:
will try to extend `users_data` to include principal

taggrx commented on src/frontend/src/wallet.tsx:
```diff
@@ -140,7 +140,7 @@ export const Wallet = () => {
                                 recipient,
                             );
-                            if (typeof response == "string")
+                            if (isNaN(+(response || "")))
```
```suggestion
                            if (Number.isNaN(Number(response || "")))
```

taggrx commented on src/frontend/src/wallet.tsx:
```diff
@@ -297,8 +297,8 @@ export const Wallet = () => {
                         );
-                        if (typeof response == "string")
-                            showPopUp("error", response);
+                        if (isNaN(+(response || "")))
```
```suggestion
                        if (Number.isNaN(Number(response || "")))
```

aligatorr89 committed:
review fixes

aligatorr89 committed:
icrc-3 blocks integration and

aligatorr89 committed:
pipeline fixes

Remove allowed tokens validation

rebase fixes

Add principal to user data, review comments

Review comments

lint fix

small fix

Fix tests

xqxpx committed:
Update src/backend/env/tip.rs

Commit:
cleanups

Commit:
ui cleanup

Commit:
popup clean up

Commit:
fix build

Commit:
clean up

Commit:
clean up types

Commit:
minor

gamer7860 committed:
disallow tipping system posts

gamer7860 committed:
disallow negative nunmbers

gamer7860 committed:
UI improvements

gamer7860 committed:
don't block on tip finalization

gamer7860 committed:
improve UI

gamer7860 committed:
Merge branch 'main' into feature-request-1413707-tips-2
